omegaconf/omegaconf.py

```diff
@@ -101,43 +100,62 @@ def SI(interpolation: str) -> Any:
 
 
 def register_default_resolvers() -> None:
-    def env(key: str, default: Any = _EMPTY_MARKER_) -> Any:
+    # DEPRECATED: remove in 2.2
+    def legacy_env(key: str, default: Optional[str] = None) -> Any:
+        warnings.warn(
+            "The `env` resolver is deprecated, see https://github.com/omry/omegaconf/issues/573",
+        )
+
         try:
-            val_str = os.environ[key]
+            return decode_primitive(os.environ[key])
         except KeyError:
-            if default is not _EMPTY_MARKER_:
-                return default
+            if default is not None:
+                return decode_primitive(default)
             else:
                 raise ValidationError(f"Environment variable '{key}' not found")
 
-        # We obtained a string from the environment variable: we try to parse it
-        # using the grammar (as if it was a resolver argument), so that expressions
-        # like numbers, booleans, lists and dictionaries can be properly evaluated.
-        try:
-            parse_tree = parse(
-                val_str, parser_rule="singleElement", lexer_mode="VALUE_MODE"
+    def env(key: str, default: Any = _EMPTY_MARKER_) -> Optional[str]:
+        if (
+            default is not _EMPTY_MARKER_
+            and default is not None
+            and not isinstance(default, str)
+        ):
+            raise ValidationError(
+                f"The default value of the `oc.env` resolver must be a string or "
+                f"None, but `{default}` is of type {type(default).__name__}"
             )
-        except GrammarParseError:
-            # Un-parsable as a resolver argument: keep the string unchanged.
-            return val_str
 
-        # Resolve the parse tree. We use an empty config for this, which means that
-        # interpolations referring to other nodes will fail.
-        empty_config = DictConfig({})
         try:
-            val = empty_config.resolve_parse_tree(parse_tree)
-        except InterpolationKeyError as exc:
-            raise InterpolationKeyError(
-                f"When attempting to resolve env variable '{key}', a node interpolation "
-                f"caused the following exception: {exc}. Node interpolations are not "
-                f"supported in environment variables: either remove them, or escape "
-                f"them to keep them as a strings."
-            ).with_traceback(sys.exc_info()[2])
+            return os.environ[key]
+        except KeyError:
+            if default is not _EMPTY_MARKER_:
+                assert default is None or isinstance(default, str)  # redundant for mypy
```

Comment:
you can probably get rid of this now.
In the future, we may use Optional[str] in the signature for type validation.

Reply:
Done in d79a4c7